# SQL SERVER

### Creating, altering and dropping database

In [ ]:
CREATE DATABASE KUDVENKAT;

In [ ]:
ALTER DATABASE KUDVENKAT MODIFY Name = KudvenkatDb

### Rename database using stored procedure...

In [ ]:
-- sp_renameDB '<oldDbName>', '<newDbName>';
sp_renameDB 'KudvenkatDb', 'KudvenkatDb_Master';


In [1]:
USE KudvenkatDb_Master;

Commands completed successfully.

Total execution time: 00:00:00.011

## Get the list of tables exist in the database...

In [3]:
SELECT table_name, table_schema, table_type
FROM information_schema.tables
ORDER BY table_name ASC;

(2 rows affected)

Total execution time: 00:00:00.156

table_name table_schema table_type tblGender dbo BASE TABLE tblPerson dbo BASE TABLE

## Drop a database...

In [ ]:
-- DROP DATABASE KudvenkatDb_Master;

## Creating and working with tables

In [ ]:
CREATE TABLE tblGender(
    ID INT NOT NULL PRIMARY KEY,
    Gender NVARCHAR(50) NOT NULL
);

In [ ]:
CREATE TABLE tblPerson(
    ID INT NOT NULL PRIMARY KEY,
    Name NVARCHAR(50) NOT NULL,
    Email NVARCHAR(50) NOT NULL,
    GenderID int NULL,
    Age int NULL
);

## Drop tables...

In [ ]:
DROP TABLE tblPerson;
DROP TABLE tblGender;

## Insert data...

In [ ]:
INSERT INTO tblGender (ID, Gender) VALUES 
            (1, 'Male'),
            (2,'Female'),
            (3, 'Unknown')

In [ ]:
INSERT INTO tblPerson (ID, Name, Email, GenderID) VALUES
            (1, 'Jade','j@j.com',2),
            (2,'Mary','m@m.com',3),
            (3,'Martin','ma@ma.com',1),
            (4,'Rob','r@r.com',NULL),
            (5,'May','may@may.com',2),
            (7,'Rich','ri@ri.com',2),
            (8,'Mike','mi@mi.com',NULL)

## Fetch data...

In [ ]:
SELECT * FROM tblGender;
SELECT * FROM tblPerson;

## Truncate tables...

In [ ]:
TRUNCATE TABLE tblPerson;

In [ ]:
TRUNCATE TABLE tblGender;

## Create Foreign Key - constraint

In [ ]:
ALTER TABLE tblPerson ADD CONSTRAINT tblPerson_GenderID_FK FOREIGN KEY (GenderID) REFERENCES tblGender(ID);

## Create default constraint

In [ ]:
-- if value for the GenderID column is not provided then insert the default value "3" in the GenderID
ALTER TABLE tblPerson ADD CONSTRAINT DF_tblPerson_GenderID DEFAULT 3 FOR GenderID;

## Drop Constraint

In [ ]:
-- ALTER table <table_name> DROP <CONSTRAINT_NAME>;

## Cascading referential integrity
* This constraint allows to define the actions MicrosoftSQL Server should take when a user attemps to delte or update a key to which an existing foreign keys points.
-  For example, if you delete row with ID =1 from tblGender table, then row with ID = 3 form tblPerson table becomens an opphan record. You will not be able to tell the Gender for this row. So, Cascading referential integrity constraint can be used to define actions MicrosoftSQL Server should take when this happerns. By default, we get an error and the DELETE or UPDATE statement is rolled back.

### Options when setting up Cascading referential integrity constraint:
    1. No Action: This is the defautl behaviour. It specifies that if an attempt is amde to delete or update a row with the key referenced by foreign keys in existing rows in other tables, an error is raised and the DELETE or UPDATE is rolled back...
    2. Cascade: Specifies that if an attempt is made to delete or update a row with a key referenced by foreign keys in existing rows in other tables, all rows containing those foreign keys are also deleted or updated...
    3. Set NULL: Specifies that if an attempt is made to delete or update a row with a key referenced by foreign keys in existing rows in other tables, all rows containing those foreign keys are set to NULL...
    4. Set Default: Specifies that if an attempt is made to delete or update a row with a key referenced by foreign keys in existing rows in other tables, all rows containing those foreign keys are set to default values... 

In [ ]:
ALTER TABLE tblPerson ADD CONSTRAINT tblPerson_GenderID_FK FOREIGN KEY (GenderID) REFERENCES tblGender(ID) ON DELETE CASCADE ON UPDATE SET NULL;

## Check constraint
    * It is used to limit the range of the values, that can be entered for a column...
    * If the column is nullable, It is possible to pass null for the column, when you will pass the value as NULL, the boolean expression evaluates to UNKNOWN, and allows the value i.e., then it will not give an error because NULL will be cosnidered as unknown, hence It will insert the NULL VALUE...

In [ ]:
-- ALTER TABLE <table_name> ADD CONSTRAINT <CONSTRAINT_NAME> CHECK <BOOLEAN_EXPRESSION
ALTER Table tblPerson ADD CONSTRAINT Chk_tblPerson_Age CHECK (AGE > 0 AND AGE < 150);

## Identity Column
- IDENTITY(increment_seed, increment_value)
- If a column is marked as an identity column, then the values for this column are automatically generated, when you insert a new row into the table...

- To Explicitly supply a value for identity column:
    1. First turn on identity insert -> SET identity_insert <table_name> ON
    2. In the insert query specifiy the column list - Insert into table_name (col1, col2,....) Values (Val1,Val2,...)

- If you have deleted all the rows in a table, and you want to reset the identity column value, user "DBCC CHECKIDENT('<table_name>',RESEED,0)

| DBCC - Database Consistancy Check

## Get the last generated identity column value in SQL Server

    Two ways to  achieve the above:
    1. The most common wat is to use SCOPE_IDENTITY() built in function...
    2. You can use @@IDENTITY
    3. IDENT_CURRENT('<table_name>')

    DIFFERENCE
    - SCOPE_IDENTITY() -> Same session and the same scope
    - @@IDENTITY -> Same session and across any scope
    - IDENT_CURRENT('<table_name>') -> Specifies table across any session and any scope

In [ ]:
CREATE TABLE Test1(
    Id int IDENTITY PRIMARY KEY,
    Name NVARCHAR(50)
);

CREATE TABLE Test2(
    Id int IDENTITY PRIMARY KEY,
    Name NVARCHAR(50)
);



In [ ]:
SELECT * FROM Test1;
SELECT * FROM Test2;

In [ ]:
INSERT INTO Test1 VALUES ('AAAAA1')
INSERT INTO Test2 VALUES ('AAAAA2')

In [ ]:
SELECT SCOPE_IDENTITY();
SELECT @@IDENTITY;
SELECT IDENT_CURRENT('Test2');

## Unique key constraint
    - We use UNIQUE constraint to enforce uniqueness of a column i.e., the column shouldn't allow any duplicate values. We can add a Unique constraint through the designer or using a query.
    
    - To create the unique key using a query:
        Alter Table <Table_name>
        Add Constraint <Constraint_name> Unique(Column_name);

    DIFFERENCE B/W PRIMARY KEY AND UNIQUE KEY CONSTRAINT
    1. A table can can have only one primary key, but more than one Unique key
    2. Primary key does not allow nulls, where as unique key allows one null

In [ ]:
ALTER TABLE tblPerson
ADD CONSTRAINT UK_tblPerson_Email UNIQUE(Email)

## Select statement

```
Operators and Wild Cards
----------------------------------------------
Operator     |     Purpose
----------------------------------------------
=            |     Equal to
!= or <>     |     Not equal to
>            |     Greater than
>=           |     Greater than or equal to
<            |     Less than 
<=           |     Less than or equal to
IN           |     Specify a list of values
BETWEEN      |     Specify a range
LIKE         |     Specify a pattern
NOT          |     Not in a list, range etc...
%            |     Specifies zero or more characters
-            |     Specifies exactly one characters
[]           |     Any character with in the brackets
[^]          |     Not any character with in the brackets

```

In [ ]:
SELECT DISTINCT Name,GenderID FROM tblPerson;
SELECT * FROM tblPerson WHERE GenderID <> 1;
SELECT * FROM tblPerson WHERE GenderID IN(2,3,5);

-- Select all the records whose name starts with any specific character
SELECT * FROM tblPerson WHERE Name LIKE 'R%';

-- Email validation... which contains @ in the email
SELECT * FROM tblPerson WHERE Email LIKE '%@%';

-- Fetch all the records where email contains 1 character before @ and one character after @
SELECT * FROM tblPerson WHERE Email LIKE '_@_.com';

-- Select the records where name starts with the Either M, S, or T
SELECT * FROM tblPerson WHERE Name LIKE '[MST]%';


In [ ]:

-- Order by for sorting | default order is ascending
SELECT * FROM tblPerson ORDER BY Name DESC, Age ASC;

SELECT TOP 10 * FROM tblPerson;

-- Select the record whose age is oldest...
SELECT TOP 1 * FROM tblPerson ORDER BY Age DESC;

## Group by in SQL Server

- Group by clause is used to group a selected set of rows into a set of summary rows by the values of one or more columns or expressions. 
- It is always used in conjunction with one or more aggregate functions.

    ### Having clause have to after group by...

## Where VS Having
     - WHERE clause is used to filter rows before aggregation, Where as HAVING caluse is used to filter groups after aggregations.
     - WHERE clause can be used with - Select, Insert, and Update statemetns, where as HAVING clause can only be used with the Select statement.
     - WHERE filters rows before aggregation (GROUPING), where as, HAVING filters groups, after the aggregations are performed
     - Aggregate functions cananot be used in the WHERE clause, unless it is in a sub query contained in a HAVING clause, whereas, aggregate functions can be used in HAVING clause

In [ ]:
CREATE TABLE tblEmployee(
    ID INT IDENTITY(100,1) PRIMARY KEY,
    Name NVARCHAR(50),
    Gender NVARCHAR(50),
    Salary FLOAT,
    City NVARCHAR(50)
);

In [ ]:
INSERT INTO tblEmployee VALUES 
                ('Tom','Male',4000,'London'),
                ('Pam','Female',3000,'New York'),
                ('John','Male',3500,'London'),
                ('Sam','Male',4500,'London'),
                ('Todd','Male',2800,'Sydney'),
                ('Ben','Male',7000,'New York'),
                ('Sara','Female',4800,'Sydney'),
                ('Valarie','Female',5500,'New York'),
                ('James','Male',6500,'London'),
                ('Russell','Male',8800,'London')

In [ ]:
SELECT * FROM tblEmployee;

In [ ]:
SELECT City,Gender, COUNT(ID) AS TotalEmployees, SUM(Salary) AS TotalSalary FROM tblEmployee GROUP BY City, Gender;

-- HAVING...
SELECT City,Gender, COUNT(ID) AS TotalEmployees, SUM(Salary) AS TotalSalary FROM tblEmployee WHERE Gender = 'Male' GROUP BY City, Gender;
SELECT City,Gender, COUNT(ID) AS TotalEmployees, SUM(Salary) AS TotalSalary FROM tblEmployee GROUP BY City, Gender HAVING Gender = 'Male';

## Joins in SQL Server

    - Joins in SQL Server are used to retrieve data from 2 or more related tables. IN general tables are realted to each other using foreign key constrainsts.

    ## In SQL Server, there are different types of JOINS
        1. INNER JOIN
        2. OUTER JOIN
        3. CROSS JOIN

     ## Outer Joins are again divided into
        1. Left Join or Left Outer Join
        2. Right Join or Right Outer Join
        3. Full Join or Full Outer Join


    ## Generic formula
        Select      ColumnList
        From        LeftTable
        JoinType    RightTable
        ON          JoinCondition

In [29]:
CREATE TABLE tblEmployee1(
    ID INT IDENTITY(101,1) PRIMARY KEY,
    Name NVARCHAR(50),
    Gender NVARCHAR(50),
    Salary FLOAT,
    DepartmentId INT NULL
);
CREATE TABLE tblDepartment(
    Id INT PRIMARY KEY,
    DepartmentName NVARCHAR(50),
    Location NVARCHAR(50),
    DepartmentHead NVARCHAR(50)
);

Commands completed successfully.

Total execution time: 00:00:00.007

In [31]:
ALTER TABLE tblEmployee1 ADD CONSTRAINT FK_tblEmp1_DeptID FOREIGN KEY(DepartmentId) REFERENCES tblDepartment;

Commands completed successfully.

Total execution time: 00:00:00.004

In [34]:
INSERT INTO tblEmployee1 Values
            ('Tom','Male',4000,1),
                ('Pam','Female',3000,3),
                ('John','Male',3500,1),
                ('Sam','Male',4500,2),
                ('Todd','Male',2800,2),
                ('Ben','Male',7000,1),
                ('Sara','Female',4800,3),
                ('Valarie','Female',5500,1),
                ('James','Male',6500,NULL),
                ('Russell','Male',8800,NULL)

INSERT INTO tblDepartment VALUEs
                (1, 'IT','London','Rick'),
                (2, 'Payroll','Delhi','Ron'),
                (3, 'HR','New York','Christie'),
                (4, 'Other Department','Sydney','Cindrella')
                

(10 rows affected)

Total execution time: 00:00:00.006

In [2]:
SELECT * FROM tblEmployee1;
SELECT * FROM tblDepartment;

(10 rows affected)

(4 rows affected)

Total execution time: 00:00:00.092

ID Name Gender Salary DepartmentId 102 Tom Male 4000 1 103 Pam Female 3000 3 104 John Male 3500 1 105 Sam Male 4500 2 106 Todd Male 2800 2 107 Ben Male 7000 1 108 Sara Female 4800 3 109 Valarie Female 5500 1 110 James Male 6500 NULL 111 Russell Male 8800 NULL

Id DepartmentName Location DepartmentHead 1 IT London Rick 2 Payroll Delhi Ron 3 HR New York Christie 4 Other Department Sydney Cindrella

## Inner Join
        ## INNER JOIN - returns only the matching rows between both the tables. Non matching rows are eliminated.

In [4]:
SELECT      t1.Name, t1.Gender, t1.Salary, t2.DepartmentName 
from        tblEmployee1 AS t1
INNER JOIN  tblDepartment AS t2
ON          t1.DepartmentId = t2.Id;

(8 rows affected)

Total execution time: 00:00:00.007

Name Gender Salary DepartmentName Tom Male 4000 IT Pam Female 3000 HR John Male 3500 IT Sam Male 4500 Payroll Todd Male 2800 Payroll Ben Male 7000 IT Sara Female 4800 HR Valarie Female 5500 IT

## Left Join
        ## LEFT JOIN - returns all the matching rows + non matching rows from the left table

In [16]:
SELECT         t1.Name, t1.Gender, t1.Salary, t2.DepartmentName 
from           tblEmployee1 AS t1
LEFT JOIN      tblDepartment AS t2
ON             t1.DepartmentId = t2.Id;

(10 rows affected)

Total execution time: 00:00:00.017

Name Gender Salary DepartmentName Tom Male 4000 IT Pam Female 3000 HR John Male 3500 IT Sam Male 4500 Payroll Todd Male 2800 Payroll Ben Male 7000 IT Sara Female 4800 HR Valarie Female 5500 IT James Male 6500 NULL Russell Male 8800 NULL

## Right Join
        ## RIGHT JOIN - returns all the matching rows + non matching rows from the right table

In [6]:
SELECT      t1.Name, t1.Gender, t1.Salary, t2.DepartmentName 
from        tblEmployee1 AS t1
RIGHT JOIN  tblDepartment AS t2
ON          t1.DepartmentId = t2.Id;

(9 rows affected)

Total execution time: 00:00:00.012

Name Gender Salary DepartmentName Tom Male 4000 IT John Male 3500 IT Ben Male 7000 IT Valarie Female 5500 IT Sam Male 4500 Payroll Todd Male 2800 Payroll Pam Female 3000 HR Sara Female 4800 HR NULL NULL NULL Other Department

## Full Outer Join
        ## FULL OUTER JOIN - returns all the rows from both the left and right tables, including the non matching rows.

In [7]:
SELECT      t1.Name, t1.Gender, t1.Salary, t2.DepartmentName 
from        tblEmployee1 AS t1
FULL JOIN   tblDepartment AS t2
ON          t1.DepartmentId = t2.Id;

(11 rows affected)

Total execution time: 00:00:00.014

Name Gender Salary DepartmentName Tom Male 4000 IT Pam Female 3000 HR John Male 3500 IT Sam Male 4500 Payroll Todd Male 2800 Payroll Ben Male 7000 IT Sara Female 4800 HR Valarie Female 5500 IT James Male 6500 NULL Russell Male 8800 NULL NULL NULL NULL Other Department

## Cross Join
        ## CROSS JOIN - produces the Cartesian product of the 2 tables involved in the join. For example, in the Employees table we have 10 rows and in the Departments we have 4 rows. So, a cross join between these 2 tables produces 40 rows.

        ## NOTE: Cross Join shouldn't have ON clause.

In [8]:
SELECT      t1.Name, t1.Gender, t1.Salary, t2.DepartmentName 
from        tblEmployee1 AS t1
CROSS JOIN  tblDepartment AS t2

(40 rows affected)

Total execution time: 00:00:00.017

Name Gender Salary DepartmentName Tom Male 4000 IT Pam Female 3000 IT John Male 3500 IT Sam Male 4500 IT Todd Male 2800 IT Ben Male 7000 IT Sara Female 4800 IT Valarie Female 5500 IT James Male 6500 IT Russell Male 8800 IT Tom Male 4000 Payroll Pam Female 3000 Payroll John Male 3500 Payroll Sam Male 4500 Payroll Todd Male 2800 Payroll Ben Male 7000 Payroll Sara Female 4800 Payroll Valarie Female 5500 Payroll James Male 6500 Payroll Russell Male 8800 Payroll Tom Male 4000 HR Pam Female 3000 HR John Male 3500 HR Sam Male 4500 HR Todd Male 2800 HR Ben Male 7000 HR Sara Female 4800 HR Valarie Female 5500 HR James Male 6500 HR Russell Male 8800 HR Tom Male 4000 Other Department Pam Female 3000 Other Department John Male 3500 Other Department Sam Male 4500 Other Department Todd Male 2800 Other Department Ben Male 7000 Other Department Sara Female 4800 Other Department Valarie Female 5500 Other Department James Male 6500 Other Department Russell Male 8800 Other Department

## Advanced or Intelligent Joins

    - Retrieve only the non matching rows from the left table
    - Retrieve only the non matching rows from the right table
    - Retrieve only the non matching rows from both the left and right table

In [19]:
-- Retrieve only the non matching rows from the left table

SELECT         t1.Name, t1.Gender, t1.Salary, t2.DepartmentName from tblEmployee1 AS t1
LEFT JOIN      tblDepartment AS t2
ON             t1.DepartmentId = t2.Id
WHERE          t1.DepartmentId IS NULL;

(2 rows affected)

Total execution time: 00:00:00.009

Name Gender Salary DepartmentName James Male 6500 NULL Russell Male 8800 NULL

In [20]:
-- Retrieve only the non matching rows from the right table

SELECT      t1.Name, t1.Gender, t1.Salary, t2.DepartmentName from tblEmployee1 AS t1
RIGHT JOIN  tblDepartment AS t2
ON          t1.DepartmentId = t2.Id
WHERE       t1.DepartmentId IS NULL;

(1 row affected)

Total execution time: 00:00:00.011

Name Gender Salary DepartmentName NULL NULL NULL Other Department

In [23]:
-- Retrieve only the non matching rows from both the left and right table
SELECT      t1.Name, t1.Gender, t1.Salary, t2.DepartmentName 
from        tblEmployee1 AS t1
FULL JOIN   tblDepartment AS t2
ON          t1.DepartmentId = t2.Id
WHERE       t1.DepartmentId IS NULL
OR          t2.Id IS NULL;

(3 rows affected)

Total execution time: 00:00:00.007

Name Gender Salary DepartmentName James Male 6500 NULL Russell Male 8800 NULL NULL NULL NULL Other Department

## Self Join
    Joining a table with itself - Self Join
    Self join can be classified as:
        - Inner Self Join
        - Outer Self Join (Left, Right and Full)
        - Cross Self Join
        
    |EmployeeID|Name|ManagerID|
        |---|---|---|
        |101|Mike|103|
        |102|Rob|101|
        |103|Todd|NULL|
        |104|Ben|101|
        |105|Sam|101|


In [24]:
CREATE TABLE tblEmployee2 (
    EmployeeID INT IDENTITY(101,1) PRIMARY KEY,
    Name NVARCHAR(50),
    ManagerID INT NULL
);

Commands completed successfully.

Total execution time: 00:00:00.008

In [ ]:
INSERT INTO tblEmployee2 
(Name, ManagerID) 
VALUES ('Mike',103), ('Rob',101), ('Todd', NULL), ('Ben',101), ('Sam', 101);

In [27]:
SELECT * FROM tblEmployee2;

(5 rows affected)

Total execution time: 00:00:00.005

EmployeeID Name ManagerID 101 Mike 103 102 Rob 101 103 Todd NULL 104 Ben 101 105 Sam 101

## Left Outer Self Join | Inner Self Join

In [32]:
-- Left Self Join
SELECT      E.Name AS Employeename, M.Name AS ManagerName
FROM        tblEmployee2 E
Left JOIN   tblEmployee2 M
ON          E.ManagerID = M.EmployeeID

-- Inner Self Join
SELECT      E.Name AS Employeename, M.Name AS ManagerName
FROM        tblEmployee2 E
INNER JOIN   tblEmployee2 M
ON          E.ManagerID = M.EmployeeID

-- Cross Self Join
SELECT      E.Name AS Employeename, M.Name AS ManagerName
FROM        tblEmployee2 E
CROSS JOIN   tblEmployee2 M

(5 rows affected)

(4 rows affected)

(25 rows affected)

Total execution time: 00:00:00.024

Employeename ManagerName Mike Todd Rob Mike Todd NULL Ben Mike Sam Mike

Employeename ManagerName Mike Todd Rob Mike Ben Mike Sam Mike

Employeename ManagerName Mike Mike Rob Mike Todd Mike Ben Mike Sam Mike Mike Rob Rob Rob Todd Rob Ben Rob Sam Rob Mike Todd Rob Todd Todd Todd Ben Todd Sam Todd Mike Ben Rob Ben Todd Ben Ben Ben Sam Ben Mike Sam Rob Sam Todd Sam Ben Sam Sam Sam

## Different ways to replace NULL values
    - ISNULL() function - ISNULL() takes 2 parameters, expression and replacement_value
    - CASE statement - CASE WHEN <Expression> THEN '' ELSE '' END
    - COALESCE() function - COALESCE() is similar to ISNULL()

In [7]:
-- ISNULL()
SELECT      E.Name as Employee, ISNULL(M.Name, 'No Manager') as Manager
from        tblEmployee2 E
LEFT JOIN   tblEmployee2 M
ON          E.ManagerID = M.EmployeeID

-- COALESCE()
SELECT      E.Name as Employee, COALESCE(M.Name, 'No Manager') as Manager
from        tblEmployee2 E
LEFT JOIN   tblEmployee2 M
ON          E.ManagerID = M.EmployeeID

-- CASE
SELECT      E.Name as Employee, CASE WHEN M.Name IS NULL THEN 'No Manager' ELSE M.Name END as Manager
from        tblEmployee2 E
LEFT JOIN   tblEmployee2 M
ON          E.ManagerID = M.EmployeeID



(5 rows affected)

(5 rows affected)

(5 rows affected)

Total execution time: 00:00:00.026

Employee Manager Mike Todd Rob Mike Todd No Manager Ben Mike Sam Mike

Employee Manager Mike Todd Rob Mike Todd No Manager Ben Mike Sam Mike

Employee Manager Mike Todd Rob Mike Todd No Manager Ben Mike Sam Mike

## COALESCE() Function 
    COALESCE() Function - Return the first NON NULL value

In [8]:
CREATE TABLE tblEmployee3 (
    Id INT PRIMARY KEY IDENTITY(101,1),
    FirstName NVARCHAR(50),
    MiddleName NVARCHAR(50),
    LastName NVARCHAR(50)
);

Commands completed successfully.

Total execution time: 00:00:00.008

In [11]:
INSERT INTO tblEmployee3 
    (FirstName, MiddleName, LastName) 
    VALUES
    ('Sam', NULL, NULL), (NULL,'Todd', 'Tanzan'), (NULL, NULL, 'Sara'), ('Ben', 'Parker', NULL), ('James', 'Nick', 'Nancy');

(5 rows affected)

Total execution time: 00:00:00.002

In [18]:
INSERT INTO tblEmployee3 
    (FirstName, MiddleName, LastName) 
    VALUES
    (NULL, NULL, NULL)
SELECT * FROM tblEmployee3

(1 row affected)

(6 rows affected)

Total execution time: 00:00:00.013

Id FirstName MiddleName LastName 101 Sam NULL NULL 102 NULL Todd Tanzan 103 NULL NULL Sara 104 Ben Parker NULL 105 James Nick Nancy 111 NULL NULL NULL

In [20]:
-- COALESCE() - returns the first NON NULL value
SELECT Id, COALESCE(FirstName, MiddleName, LastName, 'No Name') as Name from tblEmployee3

(6 rows affected)

Total execution time: 00:00:00.008

Id Name 101 Sam 102 Todd 103 Sara 104 Ben 105 James 111 No Name

## Union and Union All
    UNION and UNION ALL operators in SQL SERVER, are used to combine the result-set of two or more SELECT queries

    Note: For UNION and UNION ALL to work, the Number, Data types, and the order of the columns in the select statements should be same

    ---------------------------------------------------------

    ## Difference between UNION and UNION ALL
        1. UNION removes duplicate rows, where as UNION ALL does not
        2. UNION has to perform distinct sort to remove duplicates, which makes it less faster than UNION ALL

    Note: Estimated query execution plan - CTRL + L

    ---------------------------------------------------------

    ## Sorting results of a UNION or UNION ALL
        - ORDER BY clause should be used only on the last SELECT statement in the UNION query   

     ---------------------------------------------------------

     ## Difference between UNION and JOIN
        - UNION combines the result-set of two or more select queries into a single result-set which includes all the rows from all the queries in the union, where as JOINS, retrieve data from two or more tables based on logical relationships b/w the tables.

        In short, UNION combines rows from 2 or more tables, where JOINS combine columns from 2 or more table.

In [21]:
CREATE TABLE tblEmployee4(
    Id INT PRIMARY KEY IDENTITY(1001,1),
    Name NVARCHAR(50),
    Email NVARCHAR(50)
);

CREATE TABLE tblEmployee5(
    Id INT PRIMARY KEY IDENTITY(1001,1),
    Name NVARCHAR(50),
    Email NVARCHAR(50)
);

Commands completed successfully.

Total execution time: 00:00:00.017

In [ ]:
INSERT INTO tblEmployee4
        (Name, Email)
        VALUES
        ('Raj', 'R@R.com'), ('Sam','S@S.com')

INSERT INTO tblEmployee5
        (Name, Email)
        VALUES
        ('Ben', 'B@B.com'), ('Sam','S@S.com')



In [28]:
SELECT * FROM tblEmployee4
UNION ALL
SELECT * FROM tblEmployee5

SELECT * FROM tblEmployee4
UNION
SELECT * FROM tblEmployee5
SELECT * FROM tblEmployee

(4 rows affected)

(3 rows affected)

(10 rows affected)

Total execution time: 00:00:00.031

Id Name Email 1001 Raj R@R.com 1002 Sam S@S.com 1001 Ben B@B.com 1002 Sam S@S.com

Id Name Email 1001 Ben B@B.com 1001 Raj R@R.com 1002 Sam S@S.com

ID Name Gender Salary City 100 Tom Male 4000 London 101 Pam Female 3000 New York 102 John Male 3500 London 103 Sam Male 4500 London 104 Todd Male 2800 Sydney 105 Ben Male 7000 New York 106 Sara Female 4800 Sydney 107 Valarie Female 5500 New York 108 James Male 6500 London 109 Russell Male 8800 London

## Stored Procedures in sql server
    A stored procedure is group of T-SQL (Transact SQL) statements. If you have a situation, where you write the same query over and over again, you can save that specific query as a stored procedure and call it just by it's name.

    ## Use CREATE PROCEDURE or CREATE PROC statement to create SP
    Note: When naming user defined stored procedures, Micorsoft recommends not to use sp_ as a prefix. All system stored procedures, are prefixed with sp_. This avoids any ambiguity b/w user defined and system stored procedures and any conflicts, with some future system procedure.

    ## To execute the stored procedure
        1. spName
        2. EXEC spName
        3. EXECUTE spName

        Note: You can also right click on the procedure name, in object explorer in SQL Server Management Studio and select EXECUTE STORED PROCEDURE.


    ## To change the stored procedure, use ALTER PROCEDURE statement.
    ## To delete the SP, use DROP PROC 'SPName' or DROP PROCEDURE 'SPName'
    ## To encrypt the text of the SP, use WITH ENCRYPTION option. It is not poosible to view the text of an encrypted SP.

In [29]:
-- Create stored procedure...
CREATE PROCEDURE spGetEmployees
AS
BEGIN
    SELECT Name, Gender FROM tblEmployee
END

Commands completed successfully.

Total execution time: 00:00:00.005

In [35]:
-- Execute the storedProcedure...
spGetEmployees

(10 rows affected)

Total execution time: 00:00:00.010

Name Gender Ben Male James Male John Male Pam Female Russell Male Sam Male Sara Female Todd Male Tom Male Valarie Female

### Create Stored Procedure with Parameters
    Parameters and variables have an @ prefix in their name.

    To Execute:
    EXCEUTE  spName <param1>, <param2>
    EXCEUTE  spName <@paramName1 = value>, <@paramName2 = value>

    To View the text, of the stored procedure
    1. Use system stored procedure sp_helptext 'SPName'
                        OR
    2. Right click the SP in object explorer-> Scrip Procedure as-> Create To-> New Query Editor Window

In [31]:
CREATE PROCEDURE spGetEmployeesByGenderAndCity
@Gender nvarchar(50),
@City nvarchar(50)
AS
BEGIN
    SELECT * FROM tblEmployee WHERE Gender = @Gender AND City = @City
END

Commands completed successfully.

Total execution time: 00:00:00.010

In [32]:
spGetEmployeesByGenderAndCity 'Male', 'London'

(5 rows affected)

Total execution time: 00:00:00.014

ID Name Gender Salary City 100 Tom Male 4000 London 102 John Male 3500 London 103 Sam Male 4500 London 108 James Male 6500 London 109 Russell Male 8800 London

In [33]:
sp_helptext spGetEmployees

Commands completed successfully.

Total execution time: 00:00:00.100

Text -- Create stored procedure...
 CREATE PROCEDURE spGetEmployees
 AS
 BEGIN
 SELECT Name, Gender FROM tblEmployee
 END

In [34]:
-- Alter Stored Procedure
ALTER PROCEDURE spGetEmployees
AS
BEGIN
    SELECT Name, Gender FROM tblEmployee ORDER BY Name
END

Commands completed successfully.

Total execution time: 00:00:00.008

In [36]:
-- Encrypt the text of a stored procedure
ALTER PROCEDURE spGetEmployeesByGenderAndCity
@Gender nvarchar(50),
@City nvarchar(50)
WITH ENCRYPTION
AS
BEGIN
    SELECT * FROM tblEmployee WHERE Gender = @Gender AND City = @City
END

Commands completed successfully.

Total execution time: 00:00:00.004

In [37]:
sp_helptext spGetEmployeesByGenderAndCity

The text for object 'spGetEmployeesByGenderAndCity' is encrypted.

Total execution time: 00:00:00.092